File: addons/im_livechat/models/chatbot_message.py
from odoo import fields, models


class ChatbotMessage(models.Model):
    _name = "chatbot.message"
    _description = "Chatbot Message"
    _order = "create_date desc, id desc"
    _rec_name = "discuss_channel_id"

    mail_message_id = fields.Many2one(
        comodel_name="mail.message",
        string="Related Mail Message",
    )
    discuss_channel_id = fields.Many2one(
        comodel_name="discuss.channel",
        string="Discussion Channel",
        index=True,
        required=True,
        ondelete="cascade",
    )
    script_step_id = fields.Many2one(
        comodel_name="chatbot.script.step",
        string="Chatbot Step",
        index="btree_not_null",
    )
    user_script_answer_id = fields.Many2one(
        comodel_name="chatbot.script.answer",
        string="User's answer",
        ondelete="set null",
    )
    user_raw_script_answer_id = fields.Integer(
        help="Id of the script answer. Useful for statistics when answer is deleted."
    )
    user_raw_answer = fields.Html(string="User's raw answer")

    _unique_mail_message_id = models.Constraint(
        "unique (mail_message_id)",
        "A mail.message can only be linked to a single chatbot message",
    )
    _channel_id_user_raw_script_answer_id_idx = models.Index(
        "(discuss_channel_id, user_raw_script_answer_id) WHERE user_raw_script_answer_id IS NOT NULL",
    )

File: addons/im_livechat/static/tests/mock_server/mock_models/chatbot_message.js
import { fields, models } from "@web/../tests/web_test_helpers";

export class ChatbotMessage extends models.ServerModel {
    _name = "chatbot.message";

    discuss_channel_id = fields.Many2one({ relation: "discuss.channel" });
    script_step_id = fields.Many2one({ relation: "chatbot.script.step" });
    user_script_answer_id = fields.Many2one({ relation: "chatbot.script.answer" });
    user_raw_answer = fields.Html({ sanitize: true });

    _store_chatbot_message_fields(res) {
        /** @type {import("mock_models").ChatbotScriptAnswer} */
        const ChatbotScriptAnswer = this.env["chatbot.script.answer"];
        /** @type {import("mock_models").ChatbotScriptStep} */
        const ChatbotScriptStep = this.env["chatbot.script.step"];

        res.one("script_step_id", ["step_type"], {
            value: (message) => ChatbotScriptStep.browse(message.script_step_id),
        });
        res.one("user_script_answer_id", ["name"], {
            value: (message) => ChatbotScriptAnswer.browse(message.user_script_answer_id),
        });
        res.attr(
            "user_raw_answer",
            (message) => ["markup", message.user_raw_answer] // mock: html fields must be markup-wrapped
        );
    }
}
